chore(gatsby-plugin-sharp): move image processing to a worker

wardpeet opened the pull request:
## Description
Starting implementing new jobs api. Move logic to separate worker file so it can run in parallel.

One caveat:
The queuing part is moved to the worker so multiple jobs with the same inputpath aren't merged anymore. This might make sharp a little bit slower but I don't think we should care.

- [x] add tests

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/worker.js:
```diff
@@ -0,0 +1,48 @@
+    )
+  )
+    .then(() => callback())
+    .catch(err => callback(err))
```
Let's add a test for failures to ensure that errors are propagated correctly from the worker to the calling file

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -137,44 +130,41 @@ function runJobs(
+          { name: `gatsby-plugin-sharp` }
+        )
+      })
+      .catch(() => {})
```
This should probably reject with the previous message: `Failed to process image ${job.inputPath}`

This is also wrapped in a try catch block so perhaps we can do all error handling there?

DevSide replied:
I guess you have to reject all job deferred here too

wardpeet replied:
I rewrote the pipeline :)

pieh commented:
Should `worker` be the module that handle job queue? Or should it only be responsible for executing the job? I was under the impression that worker should only need to know how to execute the job and wouldn't need to care about scheduling/queueing

pieh commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -92,31 +82,34 @@ const scheduleJob = async (
+
+        if (totalJobs === jobsFinished) {
+          bar.done()
+          totalJobs = 0
```
This is not regression, but shouldn't we also set `jobsFinished` to `0` as well here if we are resetting that state?

wardpeet commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -137,44 +130,41 @@ function runJobs(
+      pluginOptions,
+    })
+      .then(() => {
+        jobs.map(job => {
```
Thanks, I rewrote the pipeline to have a better approach

wardpeet commented:
I've cleaned up a lot of things.
1) I'm throttling all jobs so I can merge all transforms of the same file together as we did before when the queue was created inside scheduler.js. It shouldn't slow down scheduling too much and give great perf with my previous tactic (it should be as fast as the original implementation).
2) Added tests for the scheduler, we didn't have them before so :tada:

Happy to hear your thoughts

muescha commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -48,31 +62,70 @@ const scheduleJob = async (
+      },
+      { name: `gatsby-plugin-sharp` }
+    )
+    bar.total++
```
this value is later overwritten in Line 139 - so this is obsolete here?

```js
  imagesToProcess++
  bar.total = imagesToProcess
```

muescha commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -1,88 +1,134 @@
+            if (bar) {
+              job.args.operations.forEach(() => {
+                bar.tick()
+              })
```
`progress.tick` can take an amount, if also the reporter.createProgressbar has the same tick(value) then we can do:
```js
bar.tick(job.args.operations.length)
```

KyleAMathews commented on packages/gatsby-plugin-sharp/src/worker.js:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @param {WorkerInput} job
+ */
+module.exports = job =>
```
Why does the worker have a queue? My understanding of what we're building is that workers are ephemeral (cloud functions can stop working at any time) so they can die on errors, etc. so state shouldn't be stored in the worker.

wardpeet replied:
Sadly we can't move the queue outside of a worker as it needs to know how many jobs it can execute at once. If we want to move it outside of the worker, we have to add concurrency to our jobs. This is hard because concurrency might vary on cloud vs local. We could say concurrency for a job is only used locally, and Infinity is used when doing cloud. Than it's up to cloud's pubsub to handle it.
```suggestion
module.exports = job =>
```
```suggestion
module.exports = job =>
```
```suggestion
module.exports = job =>
```

I'll give an example with images. Sharp is currently set to only process 1 inputFile and as many transforms of that file as it can for performance reasons. On cloud, we technically have an infinity amount of concurrency. When we put the concurrency in the worker we fix this problem as the worker knows how many it can execute at once. For local this means we only run one file at the time. For cloud, one function can run one input file at the time, which still gives us an infinity amount of parallel execution.

If we want to move it outside of the worker we'll probably want to change the api to something like
```
const queue = createJobQueue('IMAGE_PROCESSING', { concurrency: 1 })
queue .enqueue(args)
```
As in my opinion `enqueueJob('IMAGE_PROCESSING', args, { concurrency: 1})` is weird as each job potentially could change concurrency.

Does this make sense?

wardpeet replied:
Sadly we can't move the queue outside of a worker as it needs to know how many jobs it can execute at once. If we want to move it outside of the worker, we have to add concurrency to our jobs. This is hard because concurrency might vary on cloud vs local. We could say concurrency for a job is only used locally, and Infinity is used when doing cloud. Than it's up to cloud's pubsub to handle it.

I'll give an example with images. Sharp is currently set to only process 1 inputFile and as many transforms of that file as it can for performance reasons. On cloud, we technically have an infinity amount of concurrency. When we put the concurrency in the worker we fix this problem as the worker knows how many it can execute at once. For local this means we only run one file at the time. For cloud, one function can run one input file at the time, which still gives us an infinity amount of parallel execution.

If we want to move it outside of the worker we'll probably want to change the api to something like
```
const queue = createJobQueue('IMAGE_PROCESSING', { concurrency: 1 })
queue.enqueue(args)
```
As in my opinion `enqueueJob('IMAGE_PROCESSING', args, { concurrency: 1})` is weird as each job potentially could change concurrency.

Does this make sense?

sidharthachatterjee replied:
Resolving this for now, we'll revisit this later if necessary

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -1,88 +1,132 @@
+    toProcess.set(job.inputPath, {
+      id: jobId,
+      args: {
+        inputPath: job.inputPath,
```
We need to update this to conform to the new spec

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/index.js:
```diff
@@ -53,15 +53,12 @@ function queueImageResizing({ file, args = {}, reporter }) {
+    imgSrc
   )
-  const filePath = path.join(dirPath, imgSrc)
-  fs.ensureDirSync(dirPath)
```
We need to add this back (for a sanity check)

sidharthachatterjee replied:
```suggestion
  fs.ensureDirSync(outputDir)
```

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -1,88 +1,140 @@
+  (pluginOptions, boundActionCreators) => {
+    toProcess.forEach(job => {
+      const { task } = job
+      toProcess.delete(task.inputPaths[0])
```
Let's leave a comment for failure here?

```suggestion
      // We don't add this back even if it fails at the moment 
      toProcess.delete(task.inputPaths[0])
```

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -1,88 +1,140 @@
-  if (!isQueued) {
-    // Create image job
+  // Check if the output file already exists so we don't redo work.
+  if (existsSync(job.outputPath)) {
```
Add a TODO to remove once we hoist the check for if a job needs to be run or not to the job manager

```suggestion
  // TODO: Remove this once we hoist the check for if a job needs to be run or not to the job manager
  if (existsSync(job.outputPath)) {
```

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/scheduler.js:
```diff
@@ -1,88 +1,140 @@
-  if (existsSync(job.outputPath)) {
-    return { existingPromise: Promise.resolve(job) }
-  }
+  imagesFinished += job.task.args.operations.length
```
Nit: Maybe rename `imagesFinished` to `imageCount`?

sidharthachatterjee commented on packages/gatsby-plugin-sharp/src/__tests__/scheduler.js:
```diff
@@ -1,20 +1,217 @@
+      .mockImplementationOnce((cb, args) => cb(...args))
+    jest.doMock(`lodash`, () => {
+      return {
+        throttle: cb => (...args) => throttledFn(cb, args),
```
hacky boi

sidharthachatterjee commented:
Published in `gatsby-plugin-sharp@2.3.0`
